rfd/0214-identifier-first-login.md

```diff
@@ -0,0 +1,193 @@
+can match usernames. When a user visits the login page, they will be prompted to provide a username,
+after which they'll be shown all the auth connectors configured to match it.
+
+If this field is not set, the auth connector will default to matching all users.
+
+It should be noted that this username does not represent anything beyond being a reference to map to
+connectors on the Teleport side, and users can enter any arbitrary string. For example, if a user
+enters `joe@foo.com` as their username and are mapped to an Okta connector, but their Okta user is
+actually `bob@foo.com`, they will be logged into Teleport as `bob@foo.com`. This detail should be
+explicitly mentioned in the docs.
+
+### Example matchers
+
+#### Match all usernames that end in `@foo.com`
+
+```yaml
+user_matchers: [".*@foo\.com$"]
+```
+
+### Match `joe@foo.com` and all usernames that end in `@goteleport.com`
+
+```yaml
+user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+```
+
+### Example auth connector configuration:
+
+```yaml
+kind: saml
+metadata:
+  name: testconnector
+  revision: 60443273-15b2-4e5b-a2e2-9338439eac37
+spec:
+  acs: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  attributes_to_roles:
+    - name: groups
+      roles:
+        - editor
+      value: okta-admin
+    - name: groups
+      roles:
+        - auditor
+      value: okta-auditor
+    - name: groups
+      roles:
+        - access
+      value: okta-dev
+  audience: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+  cert: ''
+  display: Okta
+  entity_descriptor: |
+    <?xml version="1.0" encoding="UTF-8"?>
+    <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://www.example.com/00000000000000000000">
+    ...
+    </md:EntityDescriptor>
+  entity_descriptor_url: ''
+  issuer: http://www.example.com/00000000000000000000
+  service_provider_issuer: https://teleport.com/v1/webapi/saml/acs/testconnector
+  signing_key_pair:
+    cert: |
+      -----BEGIN CERTIFICATE-----
+      ...
+      -----END CERTIFICATE-----
+    private_key: |
+      -----BEGIN RSA PRIVATE KEY-----
+      ...
+      -----END RSA PRIVATE KEY-----
+  sso: https://www.example.com/app/00000000000000000000/sso/saml
+version: v2
+```
+
+### Implementation
+
+Upon loading the Web UI, the server returns a web config which contains some details about the cluster
+including the Teleport edition, login methods and auth connectors (stored in `window['GRV_CONFIG']`).
+This will now also include the RegEx matchers per auth connector.
+
+All matching and filtering will be done on the client side. This is so that we can avoid having to
+introduce any additional calls to the backend once the user has provided a username. This is faster
+and more performant, and also reduces complexity since it means we won't need to add and maintain any new
+endpoints. Remembered usernames will be kept in `localStorage`.
+
```

Comment:
im not sure im a fan of storing this entire map in localstorage. 

i imagined it would be a single username field, make the request, get returned the valid auth connectors. what was the reason behind not taking this path, just managing new endpoints? new endpoints I think is viable and wouldn't be too hard for backward compat.  

we could on load, fetch the endpoint and if its 404, just show the full form until the version this lands+1, where we can just rely on it always existing.

Reply:
I maybe should have made it more clear, but we aren't storing the whole map in `localStorage`, only the username string. I just meant that the actual filtering will be done on the frontend, but it will be done after the backend gives us the matchers. Though I'm moving away from this and instead going to let the backend handle the filtering since it's not ideal to expose all matchers to the client for any attacker to easily see (since this is a public route). See discussion [here](https://github.com/gravitational/teleport/pull/54886#discussion_r2098576941)